### ISSUE TITLE
stack pointer / corruption

### ISSUE THREAD

**Issue by aloucks** [llvm-dcpu16/llvm-dcpu16#160]
The following is corrupting the data in memory.

``` c
int func(int z);
int main() {
  int x = 1;
  int y = func(2);
  return x + y;
}
int func(int z) {
  z = z * 2;
  return z;
}
```

``` dasm16
        ; .file "bug_main.c"
        .text
        ;.globl  main
        ; .align        1
:main
        SET     PUSH, J
        SET     J, SP
        SUB     SP, 0x2
        SET     [0x1+J], 0x0   ; This is evaluating to SET 0x0000, 0x0
        SET     [J], 0x1
        SET     A, 0x2
        JSR     func
        SET     [0xffff+J], A
        SET     B, [J]
        ADD     B, A
        SET     A, B
        ADD     SP, 0x2
        SET     J, POP
        SET     PC, POP

        ;.globl  func
        ; .align        1
:func
        SUB     SP, 0x4
        SET     [0x3+SP], A
        SET     B, A
        SHL     B, 0x1
        SET     [0x3+SP], B
        SET     [SP], A
        SET     A, B
        ADD     SP, 0x4
        SET     PC, POP
```

Tested with: 0x10c DevKit

**Comment by krasin** [comment]
Hm... interesting. Looks like we need to subtract SP first and assign SP to J (frame pointer) after that.

**Comment by aloucks** [comment]
Do you even need a separate register for the frame pointer now that we can access [SP+n] ? 

In the example I posted, 0000 is corrupted. In other cases, the stack can be corrupted and SET PC, POP behaves erratically.

**Comment by krasin** [comment]
Sometimes (well, almost in every function), it's important to spill the register on stack and pick up it later. If you did that, your  offsets to SP are changed, [SP+n] becomes [SP+n+1]. It would be hard to generate the code that handles that and LLVM does not do it. See more details on [Wikipedia](http://en.wikipedia.org/wiki/Call_stack#The_stack_and_frame_pointers)

**Comment by ghost** [comment]
This is a duplicate of #153 and is a result of the 16 bit char. He "simply" calculcates the offsets wrong and corrupts the memory in the result. When you set back the char to 8 bit width, the stack pointer works normal.

**Comment by Blei** [comment]
Small tip @aloucks: if you compile with `-O1` or higher, much of the stack gymnastics disappears in most cases. Does not work completely in this case, but at least the generated code doesn't corrupt the memory.

**Comment by Blei** [comment]
Fixed by #167
